qiskit/circuit/quantumcircuit.py

```diff
@@ -2063,7 +2063,7 @@ def global_phase(self, angle):
             self._global_phase = angle
         else:
             # Set the phase to the [0, 2π) interval
-            angle = float(angle)
+            angle = complex(angle).real
```

Comment:
Do we need to support complex numbers here? I'd rather not since they don't make sense for the global phase and instead make sure that we only pass in real numbers further up the stack.

If we have to do this here, could we check if the imaginary part is 0 and only then cast to a real number? Otherwise users will be able to do 
```
qc.global_phase = 1j
```
and it will silently discard the global phase.

Reply:
@Cryoris I dug a little deeper here. All the test failures were in the `opflow `tests. They were computing gradients and receiving a complex answer which created `symengine `problems. None of those tests actually used complex gradients, so it seemed simple and prudent to just return a float or complex from `ParameterExpression.gradient` depending on whether there was a non-zero imaginary.

This solves the user entry problem and the current `opflow `tests. The remaining question is, should there be any `opflow `tests that use complex gradients?

Reply:
@dlasecki will probably add some tests in his work, so far I think we're ok with the current tests.